internal/pkg/instrumentation/bpf/go.opentelemetry.io/auto/sdk/bpf/probe.bpf.c

```diff
@@ -0,0 +1,219 @@
+	}
+
+	return rc;
+}
+
+SEC("uprobe/Tracer_start")
+int uprobe_Tracer_start(struct pt_regs *ctx) {
+    struct go_iface go_context = {0};
+    get_Go_context(ctx, 2, 0, true, &go_context);
+
+	struct otel_span_t otel_span;
+    __builtin_memset(&otel_span, 0, sizeof(struct otel_span_t));
+
+    start_span_params_t params = {
+        .ctx = ctx,
+        .go_context = &go_context,
+        .psc = &otel_span.psc,
+        .sc = &otel_span.sc,
+        .get_parent_span_context_fn = NULL,
+        .get_parent_span_context_arg = NULL, // Default to new root. 
+    };
+
+    start_span(&params);
+
+	void *parent_span_context = get_argument(ctx, 5);
+	long rc = write_span_context(parent_span_context, &otel_span.psc);
+	if (rc != 0) {
+		bpf_printk("failed to write parent span context: %ld", rc);
+	}
+
+	if (!is_sampled(params.sc)) {
+		// Default SDK behaviour is to sample everything. Only set the sampled
+		// value to false when needed.
+		void *sampled_ptr_val = get_argument(ctx, 6);
+		if (sampled_ptr_val == NULL) {
+			bpf_printk("nil sampled pointer");
+		} else {
+			bool false_val = false;
+			rc = bpf_probe_write_user(sampled_ptr_val, &false_val, sizeof(bool));
+			if (rc != 0) {
+				bpf_printk("bpf_probe_write_user: failed to write sampled value: %ld", rc);
+			} else {
+				bpf_printk("wrote sampled value");
+			}
+		}
+	}
+
+	void *span_context_ptr_val = get_argument(ctx, 7);
+	rc = write_span_context(span_context_ptr_val, &otel_span.sc);
+	if (rc != 0) {
+		bpf_printk("failed to write span context: %ld", rc);
+	}
+
+	void *span_ptr_val = get_argument(ctx, 4);
+    bpf_map_update_elem(&active_spans_by_span_ptr, &span_ptr_val, &otel_span, 0);
+    start_tracking_span(go_context.data, &otel_span.sc);
+
+	return 0;
+}
+
+SEC("uprobe/Span_ended")
+int uprobe_Span_ended(struct pt_regs *ctx) {
+    void *span_ptr = get_argument(ctx, 1);
+    struct otel_span_t *span = bpf_map_lookup_elem(&active_spans_by_span_ptr, &span_ptr);
+    if (span == NULL) {
+        return 0;
+    }
+    stop_tracking_span(&span->sc, &span->psc);
+    bpf_map_delete_elem(&active_spans_by_span_ptr, &span_ptr);
+
+	u64 len = (u64)get_argument(ctx, 3);
+	if (len > MAX_SIZE) {
+		bpf_printk("span data too large: %d", len);
+		return -1;
+	}
+	if (len == 0) {
+		bpf_printk("empty span data");
+		return 0;
+	}
+
+	void *data_ptr = get_argument(ctx, 2);
+	if (data_ptr == NULL) {
+		bpf_printk("empty span data");
+		return 0;
+	}
+
+	u32 key = 0;
+	struct event_t *event = bpf_map_lookup_elem(&new_event, &key);
+	if (event == NULL) {
+		bpf_printk("failed to initialize new event");
+		return -2;
+	}
+	event->size = (u32)len;
+
+	if (event->size < MAX_SIZE) {
+		long rc = bpf_probe_read(&event->data, event->size, data_ptr);
+		if (rc < 0) {
+			bpf_printk("failed to read encoded span data");
+			return -3;
+		}
+	} else {
+		bpf_printk("read too large: %d", event->size);
+		return -4;
+	}
+
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, event, sizeof(*event));
```

Comment:
We better use the common `output_span_event` function which all the other probes use. That function checks if the span is sampled or not and outputs it accordingly.
When including `span_output.h` the definition of the perf buffer is also included so the one in this file will need to be removed.
edit: if we avoid putting un-sampled spans in the `active_spans_by_span_ptr` map, we should not reach here for un-sampled spans, but it is still preferred to use that function for consistency.

Reply:
That function uses a span type definitions as the transport payload. This uses a payload of the encoded binary span data from Go. I don't think `output_span_event` is a compatible replacement, right?

Reply:
We need to pass the span context which we have here in `otel_span_t`, and that function accepts a pointer and a size which I think we can provide here as well. Does that make sense or am I missing something?

Reply:
Oh, my mistake. I was looking at the wrong function definition. Yeah, I think that makes sense, I'll update.

Reply:
Updated. PTAL.

Reply:
This was reverted to accommodate https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/1107#discussion_r1768687578. See https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/1107#discussion_r1769102558.